compiler/modules/front.syntax/parse.bal

```diff
@@ -7,20 +7,44 @@ public type FilePath record {|
     string? directory = ();
 |};
 
-public function parseModulePart(string[] lines, FilePath path, int partIndex) returns ModulePart|err:Syntax {
+public function toParser(string[] lines, FilePath path, int partIndex) returns Parser {
     SourceFile file = createSourceFile(path, lines);
-    Tokenizer tok = new (file);
-    check tok.advance();
-    ModulePart part = {
-        file,
-        partIndex,
-        defns: [],
-        importDecl: check parseImportDecl(tok)
-    };
-    while tok.current() != () {
-        part.defns.push(check parseModuleDecl(tok, part));
+    ModulePart part = { file, partIndex, defns: [], importDecl: () };
+    return new(part);
+}
+
+public class Parser {
+    private Tokenizer? tok = ();
+    private ModulePart part;
+
+    function init(ModulePart part) {
+        self.part = part;
+    }
+
+    public function parseImports() returns ImportDecl[]|err:Syntax {
+        Tokenizer tok = new (self.part.file);
+        self.tok = tok;
+        check tok.advance();
+
+        // xxx #443 should provide us with an array, replace below with that
+        ImportDecl? i = check parseImportDecl(tok);
+        self.part.importDecl = i;
+        ImportDecl[] imports = i == () ? [] : [i];
+        return imports;
+    }
+
+    public function parse() returns ModulePart|err:Syntax {
+        Tokenizer? t = self.tok;
+        if t is () {
+            panic error("parseImports() must be called before parse()");
+        }
+
+        Tokenizer tok = <Tokenizer>t;
```

Comment:
Is there a better way do this?

Reply:
Move parseImports back out of Parser. Call parseImports before you construct the parser. Construct the parser with a non-nil Tokenizer and a ModulePart that contains the list of imports. Provide a function getImports() that gets the imports from the module part.

Parser is not quite the right word. It's more like PartiallyParsedModule, but that's too long. Perhaps you can think of a better word.

Reply:
I think it's maybe not too difficult to make this readonly.

You can save Tokenizer state just by doing cloneReadOnly().  Then you just need to add a method on Tokenizer `function restore(Tokenizer t)` which resets the non-final fields from `t` (panicking if the final fields are not `===`). I expect we will need something like this for #412.

Reply:
Actually there's no need for getImports. Just don't make `part` private. We are not trying to encapsulate parsing in this class.

Reply:
Should be `getImportDecls`

Reply:
>You can save Tokenizer state just by doing cloneReadOnly(). 

I am not sure I understand, I though in Ballerina `cloneReadOnly()` doesn't work on objects.

Should I add methods to Tokenizer to dump state `function state() TokenizerState` and restore form it `function restore(TokenizerState s)`?

Reply:
To clarify:

```ballerina
    Tokenizer tok = new (file);
    readonly & Tokenizer t = tok.cloneReadOnly()
```
gives
```
undefined method 'cloneReadOnly' in object 'wso2/nballerina.front.syntax:0.1.0:Tokenizer'
```

Reply:
For the record, we decided to add a method to save state.